### PR TITLE
DB-9575 Remove forced regionLocation refresh from AbstractSMInputFormat.getSplits

### PR DESCRIPTION
**Pull request by ipraznik-splice**

### CODE REVIEW

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED

File: hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java [hunk 1]
```diff
@@ -191,6 +192,39 @@
         }
     }
 
+    /**
+     * Checks the sequence of split rows. If any gap is found, the splitting has to be recalculated again.
+     *
+     * @param lss
+     * @return
+     */
+    protected boolean isRefreshNeeded(List<InputSplit> lss) {
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java:201]: Can you add a check for the scan start/stop keys? Making sure there's no gap between the scan start key and the first split, or between the last split and the scan stop key. This check should also be made when there's only one split.

File: hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java [hunk 2]
```diff
@@ -191,6 +192,39 @@
...
+        if (lss.size() < 2 || !lss.stream().allMatch(is -> (is instanceof SMSplit))) {
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java:202]: Do we ever get something other than SMSplit? Is it just defensive programming?

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/990/

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/991/

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by ipraznik-splice** [comment]
> can we have a test for this?

SMInputFormatIT is enabled again and extended.

File: hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java (hunk 1 above)
Comment (ipraznik-splice) [reply hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java]: More checks are added.

File: hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java (hunk 2 above)
Comment (ipraznik-splice) [reply hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java]: I've changed it to assert the instanceof

File: hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java [hunk 3]
```diff
@@ -191,6 +192,66 @@
...
+                byte currentStartRow[] = ((SMSplit) inputSplits.get(i)).split.getStartRow();
+                byte prevEndRow[] = ((SMSplit) inputSplits.get(i - 1)).split.getEndRow();
+                if (compareRows(currentStartRow, prevEndRow) != 0) {
+                    LOG.warn("The gap in splits is found: current split [" + inputSplits.get(i) + "], previous split [" + inputSplits.get(i - 1) + "]");
+                    return true;
+                }
+            }
+            SMSplit firstSplit = (SMSplit) inputSplits.get(0);
+            SMSplit lastSplit = (SMSplit) inputSplits.get(inputSplits.size() - 1);
+            if (compareRows(firstSplit.split.getStartRow(), scanStartRow) != 0 || compareRows(lastSplit.split.getEndRow(), scanStopRow) != 0) {
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java:248]: This check is not correct, we can have a scan with start key = 1000 and the first partition might have key = 500.

We have to check that firstSplit.split.getStartRow() <= scanStartRow and that lastSplit.split.getEndRow() >= scanStopRow (or lastSplit.split.getEndRow().isEmpty())

Comment (ipraznik-splice) [reply hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java]: it has been corrected

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**